core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -606,10 +608,10 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
 
   }
 
-  override def maxNumConcurrentTasks(): Int = synchronized {
-    executorDataMap.values.map { executor =>
-      executor.totalCores / scheduler.CPUS_PER_TASK
-    }.sum
+  override def maxNumConcurrentTasks(rp: ResourceProfile): Int = synchronized {
+    val cpusPerTask = rp.getTaskCpus.getOrElse(scheduler.CPUS_PER_TASK)
+    val executorsWithResourceProfile = executorDataMap.values.filter(_.resourceProfileId == rp.id)
```

Comment:
I assume that could be the case or that executor number went to 0 if some died, the sum will just return 0. there is no change in logic for that.  Is there something specific you are wondering about?

Reply:
Let's say we're in a dynamic cluster and there's only minimum living executors. And user has a barrier job like:

```
rdd.withResources(rp).barrirer().mapPartition { part =>
  // do some barrier stuff
}.collect()
```

In dynamic mode, since executors are only going to launch after `SparkListenerStageSubmitted` posted. So, at the time we're `createShuffleMapStage`/`mergeResourceProfilesForStage`/`checkBarrierStageWithNumSlots`, there's no executors launched for it, right?

Reply:
Yes, executors need not have launched, launched/running executors might have failed (after computing `rdd` in the example above), or number of executors is < number of partitions in rdd.
`maxNumConcurrentTasks` will tell us how many (max) can be run based on current state.

Reply:
Actually, I was not talking about some error cases here. Here, I'm worrying about the normal procedure. And I realize now we already have precaution for it, see: https://github.com/apache/spark/blob/b29829e2abdebdf6fa9dd0a4a4cf4c9d676ee82d/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L983-L1004